upstairs/src/volume.rs

```diff
@@ -4310,4 +4437,604 @@ mod test {
+                        extent_count: 32,
+                        gen: 2,
+                        opts: CrucibleOpts {
+                            id: Uuid::new_v4(),
+                            target: vec![
+                                "[fd00:1122:3344:103::7]:19002"
+                                    .parse()
+                                    .unwrap(),
+                                "[fd00:1122:3344:101::7]:19002"
+                                    .parse()
+                                    .unwrap(),
+                                "[fd00:1122:3344:102::8]:19002"
+                                    .parse()
+                                    .unwrap(),
+                            ],
+                            ..Default::default()
+                        },
+                    }],
+                    read_only_parent: None,
+                },
+            )),
+        };
+
+        // Replace one of the ROP subvolume targets, and bump the gen of the sub
+        // volume region.
+        let mut replacement = original.clone();
+        match &mut replacement {
+            VolumeConstructionRequest::Volume {
+                sub_volumes,
+                read_only_parent,
+                ..
+            } => {
+                match &mut sub_volumes[0] {
+                    VolumeConstructionRequest::Region { gen, .. } => {
+                        *gen += 1;
+                    }
+
+                    _ => {
+                        panic!("how?!");
+                    }
+                }
+
+                let Some(read_only_parent) = read_only_parent.as_mut() else {
+                    panic!("how?!");
+                };
+
+                match read_only_parent.as_mut() {
+                    VolumeConstructionRequest::Volume {
+                        sub_volumes, ..
+                    } => match &mut sub_volumes[0] {
+                        VolumeConstructionRequest::Region { opts, .. } => {
+                            opts.target[1] = "[fd00:1122:3344:111::a]:20000"
+                                .parse()
+                                .unwrap();
+                        }
+
+                        _ => {
+                            panic!("how?!");
+                        }
+                    },
+
+                    _ => {
+                        panic!("how?!");
+                    }
+                }
+            }
+
+            _ => {
+                panic!("how?!");
+            }
+        }
+
+        let log = csl();
+        let result =
+            Volume::compare_vcr_for_update(original, replacement, &log)
+                .unwrap();
+        assert_eq!(
+            result,
+            Some((
+                "[fd00:1122:3344:101::7]:19002".parse().unwrap(),
+                "[fd00:1122:3344:111::a]:20000".parse().unwrap(),
+            )),
+        );
+    }
+
+    /// Test that changes under a second read-only parent work
+    #[test]
```

Comment:
Is this situation one that can exist in the current Nexus?

Reply:
Yeah, with sufficient disk + snapshot layering:

1. create blank disk 1
2. create snapshot A from disk 1
3. create disk 2 from snapshot A
4. create snapshot B from disk 2
5. create disk 3 from snapshot B

Disk 3 will have this deeper read-only parent